### PR TITLE
decode header value with utf-8

### PR DESCRIPTION
**Pull request by zuisong**
- [x] add test case 

---
For reference, this behavior is the same as the redirection behavior of reqwest.

see https://github.com/seanmonstar/reqwest/blob/29d4cff234b37065632512f002b9785700c51aa8/src/async_impl/client.rs#L2624

---

related issue: https://github.com/ducaale/xh/issues/323

### CODE REVIEW

File: src/redirect.rs [hunk 1]
```diff
@@ -51,8 +51,8 @@ fn get_next_request(mut request: Request, response: &Response) -> Option<Request
         response
             .headers()
             .get(LOCATION)
-            .and_then(|location| location.to_str().ok())
-            .and_then(|location| request.url().join(location).ok())
+            .and_then(|location| String::from_utf8(location.as_bytes().to_vec()).ok())
+            .and_then(|location| request.url().join(&location).ok())
```
Comment (blyxxyz) [inline src/redirect.rs:55]: I'm realizing now that if decoding or joining fails we should `log::warn!()` (after https://github.com/ducaale/xh/pull/371). We can pick that up in another PR though.

File: src/printer.rs [hunk 2]
```diff
@@ -336,8 +336,8 @@ impl Printer {
                 header_string.push_str(key.as_str());
             }
             header_string.push_str(": ");
-            match value.to_str() {
-                Ok(value) => header_string.push_str(value),
+            match String::from_utf8(value.as_bytes().to_vec()) {
```
Comment (blyxxyz) [inline src/printer.rs:339]: You can use https://doc.rust-lang.org/std/str/fn.from_utf8.html to go directly from `&[u8]` to `&str`

**Review by blyxxyz — COMMENTED** [review]
Nice catch!

**Comment by zuisong** [comment]
I'm wondering if we should change all calls to HeaderValue.to_str() to HeaderValue.to_utf8_str(). 

I think it's a good idea.

For example, this will allow us to download filenames that support UTF-8 encoding. 
https://github.com/ducaale/xh/blob/ca629c803f78834a32994f6ca6754cd6fc896e63/src/download.rs#L34

**Comment by blyxxyz** [comment]
Officially you're not supposed to do that so we should decide on a case-by-case basis.

For filenames we can actually pass the bytes through directly depending on the platform. But we should check what different servers and clients do.

**Comment by zuisong** [comment]
Yes, you're right. 

I decided to use UTF-8 encoding to decode the CONTENT_DISPOSITION response header, which is very useful for downloading files with Unicode names.
And Chrome implements it the same way.



before
```shell
❯ xh  httpbingo.org/response-headers 'content-disposition==attachment; filename="🥰.json"' -d
HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Connection: close
Content-Disposition: "attachment; filename=\"\xf0\x9f\xa5\xb0.json\""
Content-Length: 76
Content-Type: application/json; charset=utf-8
Date: Fri, 21 Jun 2024 16:07:28 GMT
Fly-Request-Id: 01J0XRV0MHF1VBYPJY576KHWSM-sjc
Server: Fly/ebd3372a (2024-06-19)
Via: 1.1 fly.io

Downloading 76 B to "response-headers.json-3"
Done. 76 B in 0.00108s (68.80 KiB/s)
```

after
```shell
❯ /Users/chen/.rust-target/debug/xh httpbingo.org/response-headers 'content-disposition==attachment; filename="🥰.json"' -d
HTTP/1.1 200 OK
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Connection: close
Content-Disposition: attachment; filename="🥰.json"
Content-Length: 76
Content-Type: application/json; charset=utf-8
Date: Fri, 21 Jun 2024 16:06:44 GMT
Fly-Request-Id: 01J0XRSP2WEE21V3XAHNGBFCBY-sjc
Server: Fly/ebd3372a (2024-06-19)
Via: 1.1 fly.io

Downloading 76 B to "🥰.json"
Done. 76 B in 0.00144s (51.63 KiB/s)
```

File: tests/cli.rs [hunk 3]
```diff
@@ -1657,6 +1678,90 @@ fn body_from_raw() {
...
+        .stdout(contains("curl https://exmaple.com/ -H 'hello: 你好'"))
+        .success();
+
+    get_command()
+        .args(["https://exmaple.com/", "hello:你好", "--curl-long"])
+        .assert()
+        .stdout(contains("curl https://exmaple.com/ --header 'hello: 你好'"))
+        .success();
+}
+
```
Comment (blyxxyz) [inline tests/cli.rs:1764]: Can you add these to the test cases at the bottom of `to_curl.rs` instead?

**Review by blyxxyz — APPROVED** [review]
Thank you!

File: src/printer.rs [hunk 4]
```diff
@@ -336,7 +337,7 @@ impl Printer {
                 header_string.push_str(key.as_str());
             }
             header_string.push_str(": ");
-            match value.to_str() {
+            match value.to_utf8_str() {
```
[the remark this answers is not in the document]
Comment (zuisong) [reply src/printer.rs]: done

[the remark this answers is not in the document]
Comment (zuisong) [reply src/printer.rs]: I still feel that outputting in UTF-8 is better.

As a reference, the output of curl looks like this:
```text
❯ curl  http://httpbingo.org/response-headers --url-query  hello=😀 -v
* Host httpbingo.org:80 was resolved.
* IPv6: (none)
* IPv4: 66.241.125.232
*   Trying 66.241.125.232:80...
* Connected to httpbingo.org (66.241.125.232) port 80
> GET /response-headers?hello=%f0%9f%98%80 HTTP/1.1
> Host: httpbingo.org
> User-Agent: curl/8.6.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< access-control-allow-credentials: true
< access-control-allow-origin: *
< content-type: application/json; charset=utf-8
< hello: 😀
< date: Sat, 22 Jun 2024 01:32:27 GMT
< content-length: 32
< server: Fly/ebd3372a (2024-06-19)
< via: 1.1 fly.io
< fly-request-id: 01J0YS5HC16CHAQGZWANZ140DP-nrt
< 
{
  "hello": [
    "😀"
  ]
}
* Connection #0 to host httpbingo.org left intact
```

**Pull request merged**